File: backend/app/geo.py
"""Lat/lng ↔ local metre grid helpers (DESIGN.md §3).

All game maths happens in a local equirectangular metre grid centred on the room, so
areas and distances are ordinary Euclidean numbers.
"""

from __future__ import annotations

import math

from app.protocol import Bounds

EARTH_RADIUS_M = 6_371_000.0


def metres_per_degree(latitude: float) -> tuple[float, float]:
    """Metres per degree of longitude and latitude at the given latitude."""
    per_lat = math.pi * EARTH_RADIUS_M / 180.0
    per_lng = per_lat * math.cos(math.radians(latitude))
    return per_lng, per_lat


def bounds_size_metres(bounds: Bounds) -> tuple[float, float]:
    """Width (east-west) and height (north-south) of the play area in metres."""
    mid_lat = (bounds.south + bounds.north) / 2
    per_lng, per_lat = metres_per_degree(mid_lat)
    return (bounds.east - bounds.west) * per_lng, (bounds.north - bounds.south) * per_lat


def contains(bounds: Bounds, lat: float, lng: float) -> bool:
    return bounds.south <= lat <= bounds.north and bounds.west <= lng <= bounds.east

File: frontend/lib/geo.ts
import type { Bounds } from "@/lib/protocol";

const EARTH_RADIUS_M = 6_371_000;

/** Metres per degree of longitude and of latitude at a given latitude. */
export function metresPerDegree(latitude: number): [number, number] {
  const perLat = (Math.PI * EARTH_RADIUS_M) / 180;
  return [perLat * Math.cos((latitude * Math.PI) / 180), perLat];
}

export function boundsSizeMetres(bounds: Bounds): [number, number] {
  const [perLng, perLat] = metresPerDegree((bounds.south + bounds.north) / 2);
  return [
    (bounds.east - bounds.west) * perLng,
    (bounds.north - bounds.south) * perLat,
  ];
}

export function boundsToPolygon(bounds: Bounds): GeoJSON.Feature<GeoJSON.Polygon> {
  const { south, west, north, east } = bounds;
  return {
    type: "Feature",
    properties: {},
    geometry: {
      type: "Polygon",
      coordinates: [
        [
          [west, south],
          [east, south],
          [east, north],
          [west, north],
          [west, south],
        ],
      ],
    },
  };
}

export function boundsFromCorners(
  a: { lng: number; lat: number },
  b: { lng: number; lat: number },
): Bounds {
  return {
    south: Math.min(a.lat, b.lat),
    north: Math.max(a.lat, b.lat),
    west: Math.min(a.lng, b.lng),
    east: Math.max(a.lng, b.lng),
  };
}
